File: src/services/mesh_validation.ts
/**
 * ANIGO — validação de malha renderizável (P1 — Robustez, item 3).
 *
 * Regra: **nenhum** buffer de GPU é criado a partir de uma malha que não passou
 * por aqui (GLB, malha base ou geometria canônica do núcleo). Assim uma malha
 * corrompida vira diagnóstico recuperável (`mesh_invalid`/`geometry_rejected`)
 * em vez de um `ValidationError` do WebGPU no meio de um frame — ou, pior, de um
 * buffer com lixo que só aparece como artefato visual.
 *
 * Os códigos vêm do render contract (`mesh_validation.codes`), o mesmo documento
 * que o Rust lê em `crates/anigo-renderer/src/mesh_validation.rs`.
 */
import { RENDER_CONTRACT } from "../contracts/render_contract.v1";

/** Códigos de reprovação de malha declarados no contrato. */
export const MESH_VALIDATION_CODES = [
  "EMPTY_MESH",
  "INDEX_COUNT_MISMATCH",
  "INDEX_OUT_OF_RANGE",
  "NON_FINITE_VALUE",
  "BAD_VERTEX_STRIDE",
] as const;

export type MeshValidationCode = (typeof MESH_VALIDATION_CODES)[number];

/** Stride do vértice NPR canônico (72 B) declarado no contrato. */
export const VERTEX_STRIDE_BYTES: number =
  RENDER_CONTRACT.mesh_validation?.stride_bytes ?? 72;

/** Erro de malha reprovada (estruturado, com código estável). */
export class MeshValidationError extends Error {
  readonly code: MeshValidationCode;
  readonly detail: string;
  constructor(code: MeshValidationCode, detail: string) {
    super(`[MESH ${code}] ${detail}`);
    this.name = "MeshValidationError";
    this.code = code;
    this.detail = detail;
  }
}

export interface MeshValidationResult {
  ok: boolean;
  code: MeshValidationCode | null;
  message: string;
  vertexCount: number;
  indexCount: number;
}

/** Diferenças entre a lista local e a declarada no render contract. */
export function meshValidationCatalogProblems(): string[] {
  const declared = (RENDER_CONTRACT.mesh_validation?.codes ?? []).map((entry) => entry.code);
  const problems: string[] = [];
  for (const code of MESH_VALIDATION_CODES) {
    if (!declared.includes(code)) {
      problems.push(`código '${code}' usado no validador mas ausente do render contract`);
    }
  }
  for (const code of declared) {
    if (!(MESH_VALIDATION_CODES as readonly string[]).includes(code)) {
      problems.push(`código '${code}' declarado no contrato mas desconhecido do validador`);
    }
  }
  return problems;
}

/**
 * Valida a malha **antes** de criar buffers.
 *
 * `vertices` é o buffer cru empacotado no stride canônico (72 B: posição,
 * normal, uv, cor, joints, pesos). Quando a malha já está em arrays tipados
 * por atributo (caminho do GLB), use `validateAttributeMesh`.
 */
export function validateRenderableMesh(
  packed: ArrayBufferView,
  indices: ArrayBufferView,
  options: { strideBytes?: number } = {}
): MeshValidationResult {
  const stride = options.strideBytes ?? VERTEX_STRIDE_BYTES;
  const bytes = packed.byteLength;
  const indexCount = Math.floor(indices.byteLength / 4);
  const vertexCount = bytes / stride;

  if (bytes === 0 || indices.byteLength === 0) {
    return fail("EMPTY_MESH", `malha vazia (${bytes} B de vértice, ${indices.byteLength} B de índice)`);
  }
  if (stride <= 0 || bytes % stride !== 0) {
    return fail(
      "BAD_VERTEX_STRIDE",
      `${bytes} B de vértice não é múltiplo do stride de ${stride} B`
    );
  }
  if (indices.byteLength % 4 !== 0) {
    return fail("INDEX_COUNT_MISMATCH", `${indices.byteLength} B de índice não forma u32`);
  }
  if (indexCount % 3 !== 0) {
    return fail("INDEX_COUNT_MISMATCH", `${indexCount} índices não formam triângulos`);
  }

  const indexArray =
    indices instanceof Uint32Array
      ? indices
      : new Uint32Array(indices.buffer, indices.byteOffset, indexCount);
  for (let i = 0; i < indexArray.length; i += 1) {
    if (indexArray[i] >= vertexCount) {
      return fail(
        "INDEX_OUT_OF_RANGE",
        `índice[${i}] = ${indexArray[i]} aponta para fora de ${vertexCount} vértices`
      );
    }
  }

  const floats =
    packed instanceof Float32Array
      ? packed
      : new Float32Array(packed.buffer, packed.byteOffset, Math.floor(bytes / 4));
  for (let i = 0; i < floats.length; i += 1) {
    if (!Number.isFinite(floats[i])) {
      return fail("NON_FINITE_VALUE", `float[${i}] = ${floats[i]} (NaN/Infinito)`);
    }
  }

  return {
    ok: true,
    code: null,
    message: `malha ok: ${vertexCount} vértices, ${indexCount} índices`,
    vertexCount,
    indexCount,
  };
}

/**
 * Valida a malha decodificada por atributo (caminho do GLB).
 *
 * As posições/normais precisam ser finitas: um NaN em 4070 vértices destrói o
 * depth buffer e aparece como triângulo esticado na tela, não como erro.
 */
export function validateAttributeMesh(
  positions: ArrayLike<number>,
  indices: ArrayLike<number>,
  extra: { normals?: ArrayLike<number>; uvs?: ArrayLike<number> } = {}
): MeshValidationResult {
  const vertexCount = Math.floor(positions.length / 3);
  const indexCount = indices.length;

  if (vertexCount === 0 || indexCount === 0) {
    return fail(
      "EMPTY_MESH",
      `malha vazia (${vertexCount} vértices, ${indexCount} índices)`
    );
  }
  if (positions.length % 3 !== 0) {
    return fail("BAD_VERTEX_STRIDE", `${positions.length} floats de posição não formam vec3`);
  }
  if (extra.normals && extra.normals.length !== positions.length) {
    return fail(
      "BAD_VERTEX_STRIDE",
      `${extra.normals.length} floats de normal para ${positions.length} de posição`
    );
  }
  if (extra.uvs && extra.uvs.length !== vertexCount * 2) {
    return fail(
      "BAD_VERTEX_STRIDE",
      `${extra.uvs.length} floats de uv para ${vertexCount} vértices`
    );
  }
  if (indexCount % 3 !== 0) {
    return fail("INDEX_COUNT_MISMATCH", `${indexCount} índices não formam triângulos`);
  }
  for (let i = 0; i < indexCount; i += 1) {
    if (indices[i] >= vertexCount) {
      return fail(
        "INDEX_OUT_OF_RANGE",
        `índice[${i}] = ${indices[i]} aponta para fora de ${vertexCount} vértices`
      );
    }
  }
  for (let i = 0; i < positions.length; i += 1) {
    if (!Number.isFinite(positions[i])) {
      return fail("NON_FINITE_VALUE", `posição[${i}] = ${positions[i]} (NaN/Infinito)`);
    }
  }
  if (extra.normals) {
    for (let i = 0; i < extra.normals.length; i += 1) {
      if (!Number.isFinite(extra.normals[i])) {
        return fail("NON_FINITE_VALUE", `normal[${i}] = ${extra.normals[i]} (NaN/Infinito)`);
      }
    }
  }

  return {
    ok: true,
    code: null,
    message: `malha ok: ${vertexCount} vértices, ${indexCount} índices`,
    vertexCount,
    indexCount,
  };
}

/** Igual a `validateRenderableMesh`, mas lança `MeshValidationError`. */
export function assertRenderableMesh(
  packed: ArrayBufferView,
  indices: ArrayBufferView,
  options: { strideBytes?: number } = {}
): void {
  const result = validateRenderableMesh(packed, indices, options);
  if (!result.ok && result.code) throw new MeshValidationError(result.code, result.message);
}

function fail(code: MeshValidationCode, message: string): MeshValidationResult {
  return { ok: false, code, message, vertexCount: 0, indexCount: 0 };
}

File: crates/anigo-renderer/src/mesh_validation.rs
//! ANIGO — validação de malha antes de criar buffers (P1 — Robustez, item 3).
//!
//! O headless cria VBO/IBO a cada quadro; se a malha vier corrompida (índice fora
//! do intervalo, NaN, topologia quebrada), o wgpu só reclamaria depois — com
//! erro não capturado ou artefato visual. Aqui a reprovação acontece **antes**,
//! com o mesmo vocabulário do lado TypeScript (`mesh_validation.codes` no render
//! contract).
//!
//! O invariante de layout também é checado: `size_of::<Vertex>()` precisa ser o
//! stride declarado no contrato (72 B), porque o headless faz `cast_slice` do
//! vetor de vértices direto para o buffer.

use anigo_core::mesh::{Mesh, Vertex};

use crate::render_contract;

/// Motivo da reprovação (`code` é um dos códigos do contrato).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MeshProblem {
    pub code: &'static str,
    pub detail: String,
}

impl MeshProblem {
    fn new(code: &'static str, detail: impl Into<String>) -> Self {
        Self {
            code,
            detail: detail.into(),
        }
    }

    /// Mensagem pronta para diagnóstico/telemetria.
    pub fn message(&self) -> String {
        format!("[MESH {}] {}", self.code, self.detail)
    }
}

/// Códigos de validação declarados no contrato.
pub fn codes() -> Vec<&'static str> {
    let codes = match render_contract::contract()["mesh_validation"]["codes"].as_array() {
        Some(codes) => codes,
        None => return Vec::new(),
    };
    codes
        .iter()
        .filter_map(|entry| entry["code"].as_str())
        .collect()
}

/// Stride do vértice canônico declarado no contrato (72 B).
pub fn stride_bytes() -> u64 {
    render_contract::contract()["mesh_validation"]["stride_bytes"]
        .as_u64()
        .unwrap_or(72)
}

/// Valida uma malha antes de qualquer criação de buffer.
pub fn validate_mesh(mesh: &Mesh) -> Result<(), MeshProblem> {
    let vertex_count = mesh.vertices.len();
    let index_count = mesh.indices.len();

    if vertex_count == 0 || index_count == 0 {
        return Err(MeshProblem::new(
            "EMPTY_MESH",
            format!("malha vazia ({vertex_count} vértices, {index_count} índices)"),
        ));
    }

    let stride = stride_bytes();
    if std::mem::size_of::<Vertex>() as u64 != stride {
        return Err(MeshProblem::new(
            "BAD_VERTEX_STRIDE",
            format!(
                "Vertex tem {} B mas o contrato exige {} B por vértice",
                std::mem::size_of::<Vertex>(),
                stride
            ),
        ));
    }

    if index_count % 3 != 0 {
        return Err(MeshProblem::new(
            "INDEX_COUNT_MISMATCH",
            format!("{index_count} índices não formam triângulos"),
        ));
    }

    for (position, index) in mesh.indices.iter().enumerate() {
        if *index as usize >= vertex_count {
            return Err(MeshProblem::new(
                "INDEX_OUT_OF_RANGE",
                format!("índice[{position}] = {index} aponta para fora de {vertex_count} vértices"),
            ));
        }
    }

    for (position, vertex) in mesh.vertices.iter().enumerate() {
        if !is_finite_vertex(vertex) {
            return Err(MeshProblem::new(
                "NON_FINITE_VALUE",
                format!("vértice[{position}] tem NaN/Infinito em posição, normal, uv ou pesos"),
            ));
        }
    }

    Ok(())
}

fn is_finite_vertex(vertex: &Vertex) -> bool {
    vertex.position.iter().all(|value| value.is_finite())
        && vertex.normal.iter().all(|value| value.is_finite())
        && vertex.uv.iter().all(|value| value.is_finite())
        && vertex.weights.iter().all(|value| value.is_finite())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn triangle() -> Mesh {
        let mut mesh = Mesh::new();
        mesh.vertices.push(Vertex::new([0.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0]));
        mesh.vertices.push(Vertex::new([1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [1.0, 0.0]));
        mesh.vertices.push(Vertex::new([0.0, 1.0, 0.0], [0.0, 1.0, 0.0], [0.0, 1.0]));
        mesh.indices.extend_from_slice(&[0, 1, 2]);
        mesh
    }

    #[test]
    fn codes_come_from_the_contract() {
        let codes = codes();
        assert!(codes.contains(&"EMPTY_MESH"));
        assert!(codes.contains(&"INDEX_OUT_OF_RANGE"));
        assert!(codes.contains(&"NON_FINITE_VALUE"));
        assert!(codes.contains(&"BAD_VERTEX_STRIDE"));
        assert!(codes.contains(&"INDEX_COUNT_MISMATCH"));
    }

    #[test]
    fn vertex_stride_matches_the_contract() {
        // O headless faz `cast_slice(&mesh.vertices)` direto para o VBO: se o
        // layout do `Vertex` divergir do contrato, o buffer sai torto.
        assert_eq!(std::mem::size_of::<Vertex>() as u64, stride_bytes());
    }

    #[test]
    fn valid_triangle_passes() {
        assert!(validate_mesh(&triangle()).is_ok());
    }

    #[test]
    fn empty_mesh_is_rejected() {
        let problem = validate_mesh(&Mesh::new()).expect_err("malha vazia reprovada");
        assert_eq!(problem.code, "EMPTY_MESH");
    }

    #[test]
    fn out_of_range_index_is_rejected() {
        let mut mesh = triangle();
        mesh.indices[2] = 99;
        let problem = validate_mesh(&mesh).expect_err("índice fora do intervalo");
        assert_eq!(problem.code, "INDEX_OUT_OF_RANGE");
        assert!(problem.detail.contains("99"));
    }

    #[test]
    fn broken_topology_is_rejected() {
        let mut mesh = triangle();
        mesh.indices.pop();
        let problem = validate_mesh(&mesh).expect_err("topologia quebrada");
        assert_eq!(problem.code, "INDEX_COUNT_MISMATCH");
    }

    #[test]
    fn non_finite_position_is_rejected() {
        let mut mesh = triangle();
        mesh.vertices[1].position[0] = f32::NAN;
        let problem = validate_mesh(&mesh).expect_err("NaN reprovado");
        assert_eq!(problem.code, "NON_FINITE_VALUE");
    }

    #[test]
    fn message_carries_the_code() {
        let problem = MeshProblem::new("EMPTY_MESH", "vazio");
        assert!(problem.message().contains("EMPTY_MESH"));
    }
}
